csrc/scheduler/utils.cpp

```diff
@@ -2540,6 +2541,61 @@ bool isResharding(Fusion* fusion) {
       exprs.begin(), exprs.end(), [](Expr* e) { return isResharding(e); });
 }
 
+void moveNonConcretizedBroadcastInnermost(
+    Fusion* fusion,
+    const std::unordered_set<TensorView*>& ignored_tvs) {
+  IdModel id_model(fusion);
+  const auto& exact_model = id_model.idGraph(IdMappingMode::EXACT);
+
+  ValGroups ignored_groups;
+  for (auto ignored_tv : ignored_tvs) {
+    for (auto id : ignored_tv->domain()->allIDs()) {
+      if (id->isBroadcast()) {
+        ignored_groups.pushBack(exact_model.toGroup(id));
+      }
+    }
+  }
+
+  for (auto tv : ir_utils::allTvs(fusion)) {
+    std::vector<int64_t> broadcast_to_move;
+    for (const auto i : c10::irange(tv->getLoopDomain().size())) {
+      auto loop_id = tv->getLoopDomain().at(i);
+      if (!loop_id->isBroadcast()) {
+        continue;
+      }
+
+      const auto& id_group = exact_model.toGroup(loop_id);
+      if (ignored_groups.has(id_group)) {
+        continue;
+      }
+
+      // If the exact group has a non-broadcast domain, it means it's
+      // concretized
```

Comment:
You're right. Don't know why I thought this way...

I think we want to use here the exact graph + broadcast mappings of https://github.com/NVIDIA/Fuser/blob/main/csrc/id_model/id_model.cpp#L467-L472 but without the forwardings of https://github.com/NVIDIA/Fuser/blob/main/csrc/id_model/id_model.cpp#L440-L465. I feel that the permissive graph should only include the former and the current version of the permissive graph should only be created temporarily for the loop graph. 

Or, we could add another graph, maybe simply called "Broadcast". What do you think?

Reply:
I think having a "Broadcast" graph is a good idea, for three reason.
- First, "Broadcast" is a more informative name. "Permissive" is like, "permissive on what"?
- Second, we have discussions about mathematical properties offline. "Broadcast" graph is a good citizen like exact graph that satisfy all the mathematical properties. (under the constraint that we disallow non-unique broadcast in lowering)
- Third, I also believe we should use "Broadcast" graph instead of "Permissive" graph for loop promotion, by mechanically inserting new broadcasts in loop domain. We still do not have a definitive answer for this change, however, having a separate "Broadcast" and "Permissive" graph makes it more convenient for prototyping because we can just hot-swap `s/Permissive/Broadcast/g`.

Reply:
Actually, I'm not sure due to 

> under the constraint that we disallow non-unique broadcast in lowering

That's currently enforced by the schedulers, but if all we need is just to tell if a broadcast domain is concretized, we could easily do the analysis just using the exact graph. I just can't wrap my head around the idea of the "Broadcast" graph with non-unique broadcast domains...

While non-unique broadcast domains are not something urgent for us right now, I think we could support them in the loop promotion analysis. What matters in the analysis is inlining, and since inlining implies the expansion of a broadcast domain to a concrete size, if one broadcast domain is inlined to multiple concrete domains, those concrete domains must have the same extent.

Reply:
> but if all we need is just to tell if a broadcast domain is concretized, we could easily do the analysis just using the exact graph

How? I think effectively what this PR does is to check if a broadcast is exactly mapped to the reference tensor, and if not, move it to innermost. It does not matter if this broadcast is concretized or not. I believe this is fine. It's just the tile of this PR and the name of the function needs to be changed. I can not think of a way to easily analyze if a broadcast is concretized using the exactly graph instead of "broadcast graph".

> I just can't wrap my head around the idea of the "Broadcast" graph with non-unique broadcast domains...

With our current way of how we mechanically implement broadcast, I believe "Broadcast" graph with non-unique broadcast domains is just ill defined. 

I am not suggesting that we should do anything, I am just saying that, because we currently reject non-unique broadcasts, we are safe. But we need to be really careful if we want to lift this restriction.

> I think we could support them in the loop promotion analysis.

Are you saying that the current loop promotion as-is already support this? I don't think so:

For example, if we have
<img width="1153" alt="image" src="https://github.com/user-attachments/assets/9a568346-1754-4d04-8c02-96cb68c406a1">
and I inline T3, T4, T5, T6 at 2. Will we properly handle this promotion? In the first step of loop promotion, b will be mapped to some concrete ID, but this can neither be I0 nor I1?

Reply:
> > but if all we need is just to tell if a broadcast domain is concretized, we could easily do the analysis just using the exact graph
> 
> How? 

I was thinking about something like below:

1. For each of the logical broadcast domains, mark its exact group as concretized if it's used with non-broadcast domains (https://github.com/NVIDIA/Fuser/blob/main/csrc/id_model/id_model.cpp#L467-L472).
2. Any broadcast exact group that is not marked at step 1 is not concretized. For each tensor, if it has a non-concretized logical broadcast domain, mark any of dependent loop domains as non-concretized. Usually, at this point, there should be no scheduling applied yet, so `loop_domain == logical_domain`, but it may not be the case, e.g., with multi-GPU sharding. In any case, I think it's an error if a loop domain depends on both concretized and non-concretized logical domains.

I think this covers all non-concretized broadcast domains that should be moved.

> I think effectively what this PR does is to check if a broadcast is exactly mapped to the reference tensor, and if not, move it to innermost. It does not matter if this broadcast is concretized or not. 

If concretized, it means it's "permissively" mapped with a non-broadcast domain, and since any non-broadcast domain needs to be permissively mapped with a reference tensor, it should be properly scheduled by the transform scheduler and it should not be necessary to move concretized broadcast domains.

> I believe this is fine. It's just the title of this PR and the name of the function needs to be changed. I can not think of a way to easily analyze if a broadcast is concretized using the exactly graph instead of "broadcast graph".
> 
> > I just can't wrap my head around the idea of the "Broadcast" graph with non-unique broadcast domains...
> 
> With our current way of how we mechanically implement broadcast, I believe "Broadcast" graph with non-unique broadcast domains is just ill defined.
> 
> I am not suggesting that we should do anything, I am just saying that, because we currently reject non-unique broadcasts, we are safe. But we need to be really careful if we want to lift this restriction.
> 
> > I think we could support them in the loop promotion analysis.
> 
> Are you saying that the current loop promotion as-is already support this? 

No, I don't. I just think it should be possible by some (likely non-trivial) extensions.

Reply:
> Hmm, does that mean, for example, if I my entire fusion is a set T1[I1, b] = T0[I1, b], then this means I can not schedule as I1*b?

It doesn't as long as the `b` domains of the two tensors are exactly mapped with a reference tensor.

Reply:
> We are already excluding domains mapped to reference tensor, what is the point of additionally checking if a broadcast domain is concretized or not?

Because concretized broadcast domains should be taken care by each scheduler and don't need to be moved.